New resource: azurerm_stream_analytics_reference_input_blob

praneetloke opened the pull request:
I noticed that it was not possible to create inputs for a stream analytics job of type [reference](https://docs.microsoft.com/en-us/azure/stream-analytics/stream-analytics-use-reference-data#azure-blob-storage). The code to represent such a resource already existed, but just wasn't exposed as a resource to be able to use it.

I also added an example, that I tested using my own subscription. Here's a screenshot of how the "reference" input looks.

![Screenshot_2019-06-10 Microsoft Azure](https://user-images.githubusercontent.com/1466314/59235384-c6a3d080-8ba5-11e9-90bb-f63c5c7354eb.png)

praneetloke commented on azurerm/resource_arm_stream_analytics_reference_input_blob.go:
```diff
@@ -0,0 +1,243 @@
+package azurerm
```
This code is pretty much an exact replica of the `..._stream_analytics_**stream**_input_blob.go`.

praneetloke commented on azurerm/resource_arm_stream_analytics_reference_input_blob.go:
```diff
@@ -0,0 +1,243 @@
+
+	props := streamanalytics.Input{
+		Name: utils.String(name),
+		Properties: &streamanalytics.ReferenceInputProperties{
```
The main difference between the reference_input_blob and the stream_input_blob is this `properties` block.

katbyte replied:
Would it just make sense to add this to the other resource?

praneetloke commented on azurerm/resource_arm_stream_analytics_reference_input_blob.go:
```diff
@@ -0,0 +1,243 @@
+	props := streamanalytics.Input{
+		Name: utils.String(name),
+		Properties: &streamanalytics.ReferenceInputProperties{
+			Type: streamanalytics.TypeReference,
```
This value is what sets it distinguishes a reference input from a streaming input to the stream analytics job.

praneetloke commented on azurerm/resource_arm_stream_analytics_reference_input_blob_test.go:
```diff
@@ -0,0 +1,394 @@
+package azurerm
```
The tests in here are also pretty much a mirror of the "stream"_input_blob resource's tests.

praneetloke commented on examples/stream-analytics/main.tf:
```diff
@@ -0,0 +1,61 @@
+provider "azurerm" {}
```
This is a working example. Although, I do have to clarify that I tested creating this entire plan, and also destroying it. I verified that the input for the job that is created is in fact of type "reference".

praneetloke commented on website/docs/r/stream_analytics_stream_input_blob.html.markdown:
```diff
@@ -37,8 +37,8 @@ resource "azurerm_storage_container" "example" {
 
-resource "azurerm_stream_analytics_stream_input_eventhub" "test" {
-  name                         = "eventhub-stream-input"
+resource "azurerm_stream_analytics_stream_input_blob" "test" {
```
Noticed the typos while reading the docs.

praneetloke commented:
Ping. Anyone?

katbyte commented:
As mentioned above, Would it just make sense to add this ability to the to the other steam input blob resource?

praneetloke commented:
@katbyte I am going to take another look at this and get back to you. As you can see it has been a while since I opened this PR and have to refresh my memory a bit :)

praneetloke commented:
> As mentioned above, Would it just make sense to add this ability to the to the other steam input blob resource?

Ok I looked at this again. I do believe this needs to be its own resource. Unfortunately, there is _some_ overlap in how _this_ resource is parsed vs. how the _stream_ input is parsed. But there are two places where we must explicitly use the _reference_ input type in the `*CreateUdpate` and the `*Read` functions. Maybe I don't know enough about adding a new resource that I can't see how this could be a part of the existing streaming input resource.

Just to be clear, it is important to note that the difference between the two types of input is that one if a _streaming_ input vs. what is being added with this PR is a _reference_ input. A _reference_ input is a static source of data for a Stream Analytics job, whereas a _streaming_ input is something that is continuously producing data.

Please let me know if this helps. I am happy to rework this if you have other ideas. I think for a Stream Analytics job, being able to add reference data inputs is an important feature.

praneetloke commented:
ping @katbyte

katbyte commented on azurerm/resource_arm_stream_analytics_reference_input_blob.go:
```diff
@@ -0,0 +1,243 @@
+			"path_pattern": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
```
Could we get a `validate.NoEmptyStrings` here?

praneetloke replied:
Done.

praneetloke replied:
Just curious, though. With `Required: true`, isn't `NoEmptyStrings` enforced automatically?

katbyte replied:
Nope, you can pass an empty string, alls that is required is something is assinged to the property

praneetloke replied:
Ah gotcha. Would you like me to add `validate.NoEmptyStrings` here then?

katbyte commented on azurerm/resource_arm_stream_analytics_reference_input_blob.go:
```diff
@@ -0,0 +1,243 @@
+		if _, err := client.Update(ctx, props, resourceGroup, jobName, name, ""); err != nil {
+			return fmt.Errorf("Error Updating Stream Analytics Reference Input Blob %q (Job %q / Resource Group %q): %+v", name, jobName, resourceGroup, err)
+		}
+	}
```
Given there is vastly different logic for create & update could we maybe split this function into a create and update?

katbyte commented on examples/stream-analytics/main.tf:
```diff
@@ -0,0 +1,61 @@
+}
+
+resource "azurerm_storage_container" "example" {
+  name                  = "example"
```
```suggestion
  name                  = "${var.prefix}example"
```

katbyte commented on examples/stream-analytics/main.tf:
```diff
@@ -0,0 +1,61 @@
+}
+
+resource "azurerm_stream_analytics_reference_input_blob" "test" {
+  name                         = "blob-reference-input"
```
```suggestion
  name                         = "${var.prefix}-blob-reference-input"
```

katbyte commented on examples/stream-analytics/variables.tf:
```diff
@@ -0,0 +1,7 @@
+
+variable "location" {
+  description = "The Azure Region in which all resources in this example should be created."
+}
```
Could we put a logical default location here?

praneetloke replied:
I am happy to, but I was following the other examples in this provider which don't seem to set a default location. If you would like me to still set it, I am fine with it.

katbyte commented on website/docs/r/stream_analytics_reference_input_blob.html.markdown:
```diff
@@ -0,0 +1,108 @@
+
+-> **NOTE:** This is required when `type` is set to `Csv` or `Json`.
+
+* `field_delimiter` - (Optional) The delimiter that will be used to separate comma-separated value (CSV) records. Possible values are ` ` (space), `,` (comma), `   ` (tab), `|` (pipe) and `;`.
```
```suggestion
* `field_delimiter` - (Optional) The delimiter that will be used to separate comma-separated value (CSV) records. Possible values are ` ` (space), `,` (comma), `	` (tab), `|` (pipe) and `;`.
```

katbyte commented on website/docs/r/stream_analytics_reference_input_blob.html.markdown:
```diff
@@ -0,0 +1,108 @@
+
+* `path_pattern` - (Required) The blob path pattern. Not a regular expression. It represents a pattern against which blob names will be matched to determine whether or not they should be included as input or output to the job.
+
+* `storage_account_name` - (Required) The name of the Storage Account.
```
The name of the strage account to.. ?

danielrbradley commented:
Hello, I'm personally very interested in getting Reference sources working for a project. I've read through the conversation so far and can't quite work out what the current blocker would be (beyond fixing merge conflicts). I'd be happy to try and lend a hand if I can to get this over the line and merged if you can summarise the outstanding issues that need to be resolved.

praneetloke commented:
@danielrbradley 

> beyond fixing merge conflicts

I believe this and also getting the test to pass. I haven't kept my fork up-to-date with the latest from the upstream master. I'll get this restored today.

danielrbradley commented:
@praneetloke @katbyte I've rebased and got the build working on this second PR: #5129

Quite happy @praneetloke if you'd like to pull my changes and force push up to your original branch to continue the conversation here.

praneetloke commented:
@danielrbradley I am taking a look at restoring this PR right now. Thanks for opening yours. I'll check that too. I would prefer to fix up this PR rather than opening a new one, since the comments in this PR would be easier to follow-up on.

praneetloke commented:
@katbyte @tombuildsstuff I have addressed the comments and fixed-up this PR with the latest from upstream (this repo). Can you please take another look at this?

praneetloke commented:
Almost all of the checks passed, except for the `lintstatic` target, which seems to have failed with a transient error:

```
# cd .; git clone -- https://github.com/client9/misspell /home/travis/gopath/src/github.com/client9/misspell
Cloning into '/home/travis/gopath/src/github.com/client9/misspell'...
fatal: unable to access 'https://github.com/client9/misspell/': Failed to connect to github.com port 443: Connection timed out
package github.com/client9/misspell/cmd/misspell: exit status 128
GNUmakefile:13: recipe for target 'tools' failed
```

I don't have access to restart the build. Could you please restart just [this](https://travis-ci.org/terraform-providers/terraform-provider-azurerm/jobs/624282159) job?
